File: prototype/node/src/app/core/money.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import {
  addCents,
  cents,
  centsFromColumn,
  dollarsInputValue,
  isDollarAmount,
  multiplyCents,
  negateCents,
  percentOfCents,
  formatCents,
  parseDollars,
  subtractCents,
  ZERO_CENTS,
} from './money.ts'

test('addCents sums two amounts', () => {
  assert.equal(addCents(cents(1000), cents(500)), 1500)
})

test('addCents sums two negative amounts', () => {
  assert.equal(addCents(cents(-1000), cents(-500)), -1500)
})

test('addCents sums a positive and a negative amount', () => {
  assert.equal(addCents(cents(1000), cents(-400)), 600)
})

test('cents rejects a fraction of a cent', () => {
  assert.throws(() => cents(12.5), /12\.5/)
})

test('cents rejects a non-finite amount', () => {
  assert.throws(() => cents(NaN), Error)
})

test('cents accepts a whole number of cents', () => {
  assert.equal(cents(1500), 1500)
})

test('ZERO_CENTS is nothing', () => {
  assert.equal(ZERO_CENTS, 0)
})

test('centsFromColumn reads the number a stored column returns', () => {
  assert.equal(centsFromColumn(24_900), 24_900)
})

test('centsFromColumn reads the string a wide sum returns', () => {
  assert.equal(centsFromColumn('24900'), 24_900)
})

test('centsFromColumn reads the bigint a wide sum returns', () => {
  assert.equal(centsFromColumn(24_900n), 24_900)
})

test('centsFromColumn refuses a column that is not a whole amount', () => {
  assert.throws(() => centsFromColumn('12.5'), /12\.5/)
})

test('subtractCents takes one amount off another', () => {
  assert.equal(subtractCents(cents(1500), cents(500)), 1000)
})

test('subtractCents goes below zero', () => {
  assert.equal(subtractCents(cents(500), cents(1500)), -1000)
})

test('negateCents turns an amount around', () => {
  assert.equal(negateCents(cents(1500)), -1500)
  assert.equal(negateCents(cents(-1500)), 1500)
})

test('negateCents leaves zero unsigned', () => {
  assert.equal(Object.is(negateCents(ZERO_CENTS), 0), true)
})

test('multiplyCents scales an amount by a whole quantity', () => {
  assert.equal(multiplyCents(cents(1000), 3), 3000)
})

test('multiplyCents scales a negative amount', () => {
  assert.equal(multiplyCents(cents(-1000), 3), -3000)
})

test('multiplyCents rejects a fractional factor', () => {
  assert.throws(() => multiplyCents(cents(1000), 1.5), /1\.5/)
})

test('percentOfCents takes a whole share', () => {
  assert.equal(percentOfCents(cents(1000), 10), 100)
})

test('percentOfCents rounds down below a half-cent', () => {
  assert.equal(percentOfCents(cents(333), 10), 33)
})

test('percentOfCents rounds a half-cent away from zero', () => {
  assert.equal(percentOfCents(cents(155), 10), 16)
})

test('percentOfCents rounds a negative half-cent away from zero', () => {
  assert.equal(percentOfCents(cents(-155), 10), -16)
})

test('percentOfCents rounds a half-cent away from zero at a larger amount', () => {
  assert.equal(percentOfCents(cents(1005), 10), 101)
})

test('percentOfCents rounds a negative half-cent away from zero at a larger amount', () => {
  assert.equal(percentOfCents(cents(-1005), 10), -101)
})

test('percentOfCents rejects a non-finite percent', () => {
  assert.throws(() => percentOfCents(cents(1000), NaN), Error)
})

test('formatCents writes dollars and cents', () => {
  assert.equal(formatCents(cents(1234)), '$12.34')
})

test('formatCents pads cents under a dime', () => {
  assert.equal(formatCents(cents(5)), '$0.05')
})

test('formatCents formats zero', () => {
  assert.equal(formatCents(cents(0)), '$0.00')
})

test('formatCents separates thousands', () => {
  assert.equal(formatCents(cents(1234567)), '$12,345.67')
})

test('formatCents separates millions', () => {
  assert.equal(formatCents(cents(100_000_000)), '$1,000,000.00')
})

test('formatCents puts the sign before the currency symbol', () => {
  assert.equal(formatCents(cents(-1234)), '-$12.34')
})

test('parseDollars parses dollars and cents', () => {
  assert.equal(parseDollars('12.34'), 1234)
})

test('parseDollars parses a whole dollar amount', () => {
  assert.equal(parseDollars('12'), 1200)
})

test('parseDollars parses thousands separators', () => {
  assert.equal(parseDollars('1,234.56'), 123456)
})

test('parseDollars parses a leading currency symbol', () => {
  assert.equal(parseDollars('$12.50'), 1250)
})

test('parseDollars parses a negative amount', () => {
  assert.equal(parseDollars('-12.34'), -1234)
})

test('parseDollars ignores surrounding whitespace', () => {
  assert.equal(parseDollars('  12.34  '), 1234)
})

test('parseDollars rejects an empty string', () => {
  assert.throws(() => parseDollars(''), Error)
})

test('parseDollars rejects text that is not an amount', () => {
  assert.throws(() => parseDollars('twelve dollars'), /twelve dollars/)
})

test('parseDollars rejects a fraction of a cent', () => {
  assert.throws(() => parseDollars('12.345'), /12\.345/)
})

test('parseDollars rejects more than one decimal point', () => {
  assert.throws(() => parseDollars('12.3.4'), /12\.3\.4/)
})

test('parseDollars rejects a bare minus sign', () => {
  assert.throws(() => parseDollars('-'), Error)
})

test('dollarsInputValue writes cents as a plain decimal amount', () => {
  assert.equal(dollarsInputValue(cents(45_000)), '450.00')
})

test('dollarsInputValue pads a fractional amount under ten cents', () => {
  assert.equal(dollarsInputValue(cents(105)), '1.05')
})

test('dollarsInputValue writes zero as 0.00', () => {
  assert.equal(dollarsInputValue(cents(0)), '0.00')
})

test('dollarsInputValue leaves out the thousands separators formatCents adds', () => {
  assert.equal(dollarsInputValue(cents(1_234_567)), '12345.67')
})

// Pinned, not fixed: dollarsInputValue's one call site prefills a price
// field, and a listing price is never negative. formatCents keeps the sign,
// so parseDollars(formatCents(x)) is not a round trip for a negative x, but
// nothing in the app asks dollarsInputValue to make one.
test('dollarsInputValue drops the sign of a negative amount', () => {
  assert.equal(dollarsInputValue(cents(-105)), '1.05')
})

test('isDollarAmount accepts the forms parseDollars parses', () => {
  assert.equal(isDollarAmount('249'), true)
  assert.equal(isDollarAmount('249.00'), true)
  assert.equal(isDollarAmount('$249'), true)
  assert.equal(isDollarAmount('1,234.00'), true)
  assert.equal(isDollarAmount('  12.34  '), true)
  assert.equal(isDollarAmount('-12.34'), true)
})

test('isDollarAmount refuses what parseDollars would throw on', () => {
  assert.equal(isDollarAmount('12.345'), false)
  assert.equal(isDollarAmount('free'), false)
  assert.equal(isDollarAmount(''), false)
})

File: prototype/rails/src/test/models/money_test.rb
require "test_helper"

class MoneyTest < ActiveSupport::TestCase
  test "from cents keeps the integer amount" do
    assert_equal 1234, Money.from_cents(1234).cents
  end

  test "from cents rejects a value that is not a whole number of cents" do
    assert_raises(ArgumentError) { Money.from_cents("12.34") }
  end

  test "zero is nothing, and the seed a sum of amounts starts from" do
    assert_equal 0, Money.zero.cents
    assert_equal 1500, [ Money.from_cents(500), Money.from_cents(1000) ].sum(Money.zero).cents
  end

  test "amounts with the same cents are equal" do
    assert_equal Money.from_cents(500), Money.from_cents(500)
  end

  test "from dollars parses dollars and cents" do
    assert_equal 1234, Money.from_dollars("12.34").cents
  end

  test "from dollars parses a whole dollar amount" do
    assert_equal 1200, Money.from_dollars("12").cents
  end

  test "from dollars parses thousands separators" do
    assert_equal 123456, Money.from_dollars("1,234.56").cents
  end

  test "from dollars parses a leading currency symbol" do
    assert_equal 1234, Money.from_dollars("$12.34").cents
  end

  test "from dollars parses a negative amount" do
    assert_equal(-1234, Money.from_dollars("-12.34").cents)
  end

  test "from dollars ignores surrounding whitespace" do
    assert_equal 1234, Money.from_dollars("  12.34  ").cents
  end

  test "from dollars rejects text that is not an amount" do
    assert_raises(ArgumentError) { Money.from_dollars("twelve dollars") }
  end

  test "from dollars rejects a fraction of a cent" do
    assert_raises(ArgumentError) { Money.from_dollars("12.345") }
  end

  test "addition sums the cents" do
    assert_equal Money.from_cents(1500), Money.from_cents(1000) + Money.from_cents(500)
  end

  test "multiplication scales by a count" do
    assert_equal Money.from_cents(3000), Money.from_cents(1000) * 3
  end

  test "multiplication rejects a fractional count" do
    assert_raises(ArgumentError) { Money.from_cents(1000) * 1.5 }
  end

  test "percent takes a whole share" do
    assert_equal Money.from_cents(100), Money.from_cents(1000).percent(10)
  end

  test "percent rounds a half cent away from zero" do
    assert_equal Money.from_cents(101), Money.from_cents(1005).percent(10)
  end

  test "percent rounds a negative half cent away from zero" do
    assert_equal Money.from_cents(-101), Money.from_cents(-1005).percent(10)
  end

  test "percent rounds down below a half cent" do
    assert_equal Money.from_cents(33), Money.from_cents(333).percent(10)
  end

  test "format writes dollars and cents" do
    assert_equal "$12.34", Money.from_cents(1234).format
  end

  test "format pads cents under a dime" do
    assert_equal "$0.05", Money.from_cents(5).format
  end

  test "format separates thousands" do
    assert_equal "$1,234.56", Money.from_cents(123456).format
  end

  test "format separates millions" do
    assert_equal "$1,000,000.00", Money.from_cents(100_000_000).format
  end

  test "format puts the sign before the currency symbol" do
    assert_equal "-$12.34", Money.from_cents(-1234).format
  end
end
